lib/logger/lib/logger.ex

```diff
@@ -1063,16 +1076,24 @@ defmodule Logger do
     if compare_levels(level, min_level) != :lt do
       macro_log(level, data, metadata, caller)
     else
-      no_log(data, metadata)
+      no_log_and_maybe_eval_data(data, metadata)
     end
   end
 
-  defp no_log(data, metadata) do
-    # We wrap the contents in an anonymous function
+  defp no_log_and_maybe_eval_data(data, metadata) do
+    # Either way, we wrap the contents in an anonymous function
     # to avoid unused variable warnings.
-    quote do
-      _ = fn -> {unquote(data), unquote(metadata)} end
-      :ok
+    if Application.fetch_env!(:logger, :always_evaluate_messages) do
```

Comment:
This variable is being read in compile time and you're using `fetch_env!`. Should probably use `compile_env!` (?)